File: android/runtime/src/main/kotlin/pes/store/Cloud.kt
/**
 * CloudStore abstraction (spec §8.5) and the local-folder backend. Mirrors
 * `pes/store/cloud.py`: the sync procedure (§8.4) is written against this
 * interface so the Google Drive REST backend and a plain folder
 * (development/tests) are interchangeable. Paths are always cloud-relative,
 * `/`-separated (e.g. `events/laptop-9c11aa00/2026-08.jsonl`).
 *
 * `metadata` returns an opaque change tag (`etag`): callers may only compare
 * it for equality against a previously seen value. The local backend hashes
 * content; the Drive backend uses `modifiedTime`.
 */
package pes.store

import java.io.File
import java.security.MessageDigest

data class CloudMeta(val etag: String, val size: Long)

interface CloudStore {
    /** File content, or null if absent. */
    fun get(path: String): ByteArray?

    /** Write (overwrite) a file, creating parents. */
    fun put(path: String, data: ByteArray)

    /** Write only if the file does not exist; true if written. */
    fun putIfAbsent(path: String, data: ByteArray): Boolean

    /** All file paths under a directory prefix, recursively, sorted. */
    fun list(prefix: String): List<String>

    /** Metadata, or null if absent. */
    fun metadata(path: String): CloudMeta?
}

private fun checkRelative(path: String): String {
    if (path.isEmpty()) return path // the store root itself (list(""))
    val parts = path.split("/")
    require(!path.startsWith("/") && ".." !in parts && parts.none { it.isEmpty() }) {
        "cloud path must be relative, without ..: $path"
    }
    return path
}

/** A directory on disk acting as the cloud folder. */
class LocalFolderStore(root: File) : CloudStore {
    private val root: File = root.apply { mkdirs() }

    private fun full(path: String): File = File(root, checkRelative(path))

    override fun get(path: String): ByteArray? {
        val f = full(path)
        return if (f.isFile) f.readBytes() else null
    }

    override fun put(path: String, data: ByteArray) {
        val f = full(path)
        f.parentFile?.mkdirs()
        val tmp = File(f.parentFile, f.name + ".tmp")
        tmp.writeBytes(data)
        check(tmp.renameTo(f) || (f.delete() && tmp.renameTo(f))) { "rename failed: $f" }
    }

    override fun putIfAbsent(path: String, data: ByteArray): Boolean {
        if (full(path).exists()) return false
        put(path, data)
        return true
    }

    override fun list(prefix: String): List<String> {
        val base = full(prefix)
        if (!base.isDirectory) return emptyList()
        return base.walkTopDown()
            .filter { it.isFile && !it.name.endsWith(".tmp") }
            .map { it.relativeTo(root).invariantSeparatorsPath }
            .sorted()
            .toList()
    }

    override fun metadata(path: String): CloudMeta? {
        val f = full(path)
        if (!f.isFile) return null
        val data = f.readBytes()
        val digest = MessageDigest.getInstance("SHA-256").digest(data)
        return CloudMeta(digest.joinToString("") { "%02x".format(it) }, data.size.toLong())
    }
}

File: desktop/pes/store/cloud.py
"""CloudStore abstraction (spec §8.5) and the local-folder backend.

The sync procedure (§8.4) is written against this interface so the Google
Drive REST backend (Milestone 3) and a plain shared folder (development,
Syncthing) are interchangeable. Paths are always cloud-relative, ``/``
-separated (e.g. ``events/laptop-9c11aa00/2026-08.jsonl``).

``metadata`` returns an opaque change tag (``etag``): callers may only compare
it for equality against a previously seen value. The local backend hashes
content; the Drive backend will use ``modifiedTime``.
"""

from __future__ import annotations

import hashlib
from abc import ABC, abstractmethod
from pathlib import Path, PurePosixPath


class CloudStore(ABC):
    @abstractmethod
    def get(self, path: str) -> bytes | None:
        """File content, or None if absent."""

    @abstractmethod
    def put(self, path: str, data: bytes) -> None:
        """Write (overwrite) a file, creating parents."""

    @abstractmethod
    def put_if_absent(self, path: str, data: bytes) -> bool:
        """Write only if the file does not exist; True if written."""

    @abstractmethod
    def list(self, prefix: str) -> list[str]:
        """All file paths under a directory prefix, recursively, sorted."""

    @abstractmethod
    def metadata(self, path: str) -> dict | None:
        """{"etag": str, "size": int} or None if absent."""


def _check_relative(path: str) -> PurePosixPath:
    p = PurePosixPath(path)
    if p.is_absolute() or ".." in p.parts:
        raise ValueError(f"cloud path must be relative, without ..: {path!r}")
    return p


class LocalFolderStore(CloudStore):
    """A directory on disk acting as the cloud folder."""

    def __init__(self, root: Path | str):
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)

    def _full(self, path: str) -> Path:
        return self.root / _check_relative(path)

    def get(self, path: str) -> bytes | None:
        full = self._full(path)
        if not full.is_file():
            return None
        return full.read_bytes()

    def put(self, path: str, data: bytes) -> None:
        full = self._full(path)
        full.parent.mkdir(parents=True, exist_ok=True)
        tmp = full.with_name(full.name + ".tmp")
        tmp.write_bytes(data)
        tmp.replace(full)

    def put_if_absent(self, path: str, data: bytes) -> bool:
        full = self._full(path)
        if full.exists():
            return False
        self.put(path, data)
        return True

    def list(self, prefix: str) -> list[str]:
        base = self._full(prefix)
        if not base.is_dir():
            return []
        return sorted(
            str(p.relative_to(self.root).as_posix())
            for p in base.rglob("*")
            if p.is_file() and not p.name.endswith(".tmp")
        )

    def metadata(self, path: str) -> dict | None:
        full = self._full(path)
        if not full.is_file():
            return None
        data = full.read_bytes()
        return {"etag": hashlib.sha256(data).hexdigest(), "size": len(data)}
